i#4271: Fix signal-syscall race handling on AArch64.

abhinav92003 opened the pull request:
Fixes logic to detect the two jmps preceding a syscall. These jmps are used to bypass the syscall, which bounds the time before exiting code cache and hence delivering the signal.

Also, enables the linux.signal_racesys test on AArch64. This required adding the missing implementation to check for pending signals before entering fcache.

Issue: #2043, #4271
Fixes: #4271

abhinav92003 commented:
run arm tests

abhinav92003 commented:
I was able to reproduce the original assert failure from #4271 on `linux.signal_racesys`, and fix it too. But, I'm seeing some other tests with a new failure now:
```
 Internal Error: DynamoRIO debug check failure: /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/core/unix/signal.c:6354 (app_pc)sc-SC_XIP != next_pc 
```

This comes up maybe ~half of the time on `linux.signal0111`.

Based on the comment at the assert, this is probably because of too small app stack.
https://github.com/DynamoRIO/dynamorio/blob/25ca6af6e22841d3107d161c503da1f327d56e4c/core/unix/signal.c#L6354 

@derekbruening any feedback on the current set of changes in this PR, or how they could related to this assert failure?

abhinav92003 commented on core/arch/mangle_shared.c:
```diff
@@ -741,9 +741,15 @@ mangle_syscall_code(dcontext_t *dcontext, fragment_t *f, byte *pc, bool skip)
                 /* For A32 it's not OP_b_short */
                 IF_ARM(||
                        (instr_get_opcode(&instr) == OP_jmp &&
-                        opnd_get_pc(instr_get_target(&instr)) == pc + ARM_INSTR_SIZE)))
```
I haven't verified, but could this condition be a bug for A32? It matches only when the jmp instr at `skip_pc` has `cti_pc` as target. But, it won't match when the instr at `skip_pc` has already been modified to point to the syscall.

abhinav92003 replied:
In this case, `skip_pc` will remain NULL, and `mangle_syscall_code` will return early. This may lead to unbounded signal delivery time.

derekbruening replied:
It looks like it will match when `skip_pc` points at the syscall (`pc` is the end of the cur instr, which is the exit cti, plus one more), so it will work for the `skip == true` case.  So it will go back to dispatch.  But it will never be re-linked and so we'll always exit it looks like: so it seems like a performance bug.

derekbruening replied:
Maybe you could tweak it to fix the problem?

abhinav92003 replied:
Done.

abhinav92003 replied:
Our test suite only builds ARM32, but doesn't run tests (https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#continuous-integration-ci-via-travis-and-appveyor). Should I do a local run of the ARM32 test suite on the Raspberry Pi machine we have (https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#ci-for-aarch64--aarch32)?

derekbruening replied:
Yes, that would be nice.  Yeah, this is a gap in the CI.  IIRC the ARM folks were going to set Jenkins up on that machine but it is not there today.  There are some academics targeting Android on 32-bit ARM who I was hoping would help set sthg up but that hasn't happened either.

abhinav92003 replied:
Details of manual ARM32 testing in other comment threads.

derekbruening commented:
> I was able to reproduce the original assert failure from #4271 on `linux.signal_racesys`, and fix it too. But, I'm seeing some other tests with a new failure now:
> 
> ```
>  Internal Error: DynamoRIO debug check failure: /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/core/unix/signal.c:6354 (app_pc)sc-SC_XIP != next_pc 
> ```
> 
> This comes up maybe ~half of the time on `linux.signal0111`.
> 
> Based on the comment at the assert, this is probably because of too small app stack.
> https://github.com/DynamoRIO/dynamorio/blob/25ca6af6e22841d3107d161c503da1f327d56e4c/core/unix/signal.c#L6354
> 
> @derekbruening any feedback on the current set of changes in this PR, or how they could related to this assert failure?

Not sure I understand that comment...if the stack is too small to fit the frame, the frame will overflow the stack and either hit unreadable memory like a guard page and we would then send SIGSEGV which would likely kill the process, or if there really is no guard or adjacent unreadable you'd just clobber sthg.

derekbruening commented:
> ￼ abhinav92003 requested a review from derekbruening 8 hours ago

Sorry, missed this in the thread of other notifications.  Looking now.

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -3972,10 +3972,11 @@ if (UNIX)
-    tobuild(linux.signal_race linux/signal_race.c)
-    target_link_libraries(linux.signal_race rt)
+  if (NOT APPLE AND NOT ANDROID) # Test uses Linux-specific timer code.
+    if (NOT AARCH64) # TODO i#2043: Enable this for AArch64.
```
This PR fixes #2043, right?

abhinav92003 replied:
Yup, changed the comment to point to #1569. Will mark #2043 as fixed when creating the final commit message for this PR.

derekbruening reviewed (commented):
Could we share the append_fcache_enter_prologue code?

derekbruening commented:
> >  Internal Error: DynamoRIO debug check failure: /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/core/unix/signal.c:6354 (app_pc)sc-SC_XIP != next_pc 
> 
> Not sure I understand that comment...if the stack is too small to fit the frame, the frame will overflow the stack and either hit unreadable memory like a guard page and we would then send SIGSEGV which would likely kill the process, or if there really is no guard or adjacent unreadable you'd just clobber sthg.

Trying to understand the assert: so it's making sure the passed-to-app sc->pc != fcache_return?  Confused.

abhinav92003 commented on core/arch/aarch64/emit_utils.c:
```diff
@@ -408,18 +408,40 @@ insert_load_dr_tls_base(dcontext_t *dcontext, instrlist_t *ilist, instr_t *where
+                          opnd_create_reg(DR_REG_X1)));
+    APP(ilist, XINST_CREATE_return(dcontext));
+    APP(ilist, no_signals);
+#endif
```
Actually if I use DR_REG_R2/X2 instead of DR_REG_W2, encoding fails for the load.
```
Internal Error: Failed to encode instruction: 'ldrb   +0x0384(%x5)[1byte] -> %x2'
```
I think it needs a 32-bit width reg.

abhinav92003 replied:
But sure, I can move the logic to aarchxx, with some pre-processor directives to handle the differences between ARM and AARCH64.

abhinav92003 replied:
Done.

abhinav92003 commented on core/globals.h:
```diff
@@ -912,7 +912,15 @@ struct _dcontext_t {
+ * But we need signed char, so we explicitly qualify this declaration.
+ */
+#    if defined(AARCH64) || defined(ARM)
+    signed char signals_pending; /* != 0: pending; < 0: currently handling one */
```
I tried that (commit 5328899). But it causes failure in the macOS build: https://travis-ci.com/github/DynamoRIO/dynamorio/jobs/392370251. 

I'm not sure what's causing that, and I cannot try locally either.

derekbruening replied:
Is that #4456 decode.c nondet failure?  Unrelated to any change here.  Started appearing randomly yesterday.

abhinav92003 replied:
Yes it was. I got it in two consecutive Travis runs, so I assumed it was somehow because of the data type change. I'll revert and try again. Thanks.

abhinav92003 replied:
Didn't see the failure again.

abhinav92003 commented on suite/tests/CMakeLists.txt:
```diff
@@ -3972,10 +3972,11 @@ if (UNIX)
-    tobuild(linux.signal_race linux/signal_race.c)
-    target_link_libraries(linux.signal_race rt)
+  if (NOT APPLE AND NOT ANDROID) # Test uses Linux-specific timer code.
+    if (NOT AARCH64) # TODO i#2043: Enable this for AArch64.
```
There seems to be some other issue with linux.signal_race app. It continues to hang when run under DR, even with this PR.

abhinav92003 replied:
Based on the comments on both the test sources, `signal_racesys` is the one that "adjusts (the timer syscall) to arrive immediately **before** a system call" whereas, `signal_race` "adjusts it to arrive immediately **after** the return from the system call"

`linux.signal_racesys` seemed the relevant one to reproducing the issue in #4271.

derekbruening replied:
But the comment says #2043: so that is no longer the right issue?

Does the hang go away with -unsafe_build_ldstex?

abhinav92003 replied:
> But the comment says #2043: so that is no longer the right issue?

Probably not. I'll change it to #1569.

> Does the hang go away with -unsafe_build_ldstex?

It continues to hang.

abhinav92003 commented on core/globals.h:
```diff
@@ -912,7 +912,15 @@ struct _dcontext_t {
 
     dr_where_am_i_t whereami; /* where control is at the moment */
 #ifdef UNIX
+/* On ARM based machines, char is unsigned by default.
```
Yeah, without this `signals_pending` wraps and becomes 255, instead of -1. This causes issues with the `cmp` `le` check in `append_fcache_enter_prologue`.

abhinav92003 replied:
For completeness of this comment thread:
it seems that making the `char` `signed` didn't help with the assembly `cmp le` check (failures in `linux.signal0111`). 

But it is still required for correctness of checks in C code. Note that when I made the `char` `signed`, it did fix `linux.signest`.

More details in other comment threads in the PR.

abhinav92003 commented on core/globals.h:
```diff
@@ -912,7 +912,15 @@ struct _dcontext_t {
+ * But we need signed char, so we explicitly qualify this declaration.
+ */
+#    if defined(AARCH64) || defined(ARM)
+    signed char signals_pending; /* != 0: pending; < 0: currently handling one */
```
I did some quick searches and couldn't find any other obviously incorrect usage of the default `char`. Mostly I saw buffers for filenames, commands or disassembled instrs

Some already use the explicit `signed char` where it makes sense, like:
https://github.com/DynamoRIO/dynamorio/blob/d7321a68934eb94e9746292ea4591dbf3a24e29f/core/arch/aarchxx/mangle.c#L660
https://github.com/DynamoRIO/dynamorio/blob/d7321a68934eb94e9746292ea4591dbf3a24e29f/core/ir/x86/decode_fast.c#L122

`globals_shared` typedefs a `signed char` as `sbyte` here: https://github.com/DynamoRIO/dynamorio/blob/d7321a68934eb94e9746292ea4591dbf3a24e29f/core/lib/globals_shared.h#L219
I can probably use `sbyte` instead of signed char in this PR too.

derekbruening replied:
Sure, it makes it clear it's unrelated to characters or strings.

abhinav92003 replied:
Also found #1034.

abhinav92003 commented:
> Trying to understand the assert: so it's making sure the passed-to-app sc->pc != fcache_return? Confused.

It seems so. Based on the comment there, it signifies that the inner frame was overwritten during handling of a nested signal?

I added some logs to `handle_sigreturn`
```
MYLOG: incoming sc->SC_XIP is 281473184303696,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281473184303696

MYLOG: incoming sc->SC_XIP is 281473184474844,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281473184474844

MYLOG: incoming sc->SC_XIP is 281473184467436,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281473184467436

Sending SIGUSR2
MYLOG: incoming sc->SC_XIP is 281473184314064,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281473184314064

Sending SIGUSR1
in signal handler
Got SIGUSR1
MYLOG: incoming sc->SC_XIP is 281473184226456,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281473184226456

Sending SIGRTMAX
in signal handler
Got SIGRTMAX
MYLOG: incoming sc->SC_XIP is 281473184226456,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281473184226456

// the sc->SC_XIP here seems far away from the ones in other logs
MYLOG: incoming sc->SC_XIP is 1899019016,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 1899019016

MYLOG: incoming sc->SC_XIP is 281472031181952,  sc->fault_address is 281473185361920
MYLOG BEFORE ASSERT: sc->SC_XIP is 281472031181952, next_pc is 281472031181952
```
The second-to-last `sc->SC_XIP` (or `sc->pc`) stands out -- it's much farther away from other `SC_XIP`s in the logs. In runs where the assertion is not hit, this address doesn't appear. 
gdb shows the contents at 1899019016 (or 0x7130bb08)
```
(gdb) x/10i 1899019008
   0x7130bb00 <dynamorio_clone>:	bl	0x7130b7c0 <unexpected_return>
   0x7130bb04 <dynamorio_sigreturn>:	mov	w8, #0x8b                  	// #139
   0x7130bb08 <dynamorio_sigreturn+4>:	svc	#0x0 // <-----
   0x7130bb0c <dynamorio_sigreturn+8>:	bl	0x7130b7c0 <unexpected_return>
   0x7130bb10 <dynamorio_sys_exit>:	mov	w0, #0x0                   	// #0
   0x7130bb14 <dynamorio_sys_exit+4>:	mov	w8, #0x5d                  	// #93
   0x7130bb18 <dynamorio_sys_exit+8>:	svc	#0x0
   0x7130bb1c <dynamorio_sys_exit+12>:	bl	0x7130b7c0 <unexpected_return>
```
Trying to find out what that signifies.


Also, there's an NYI comment where `SC_XIP` is defined. Not sure what that's for.
https://github.com/DynamoRIO/dynamorio/blob/d7321a68934eb94e9746292ea4591dbf3a24e29f/core/unix/os_public.h#L137

derekbruening commented:
So a signal came in while executing dynamorio_sigreturn?  What are the signal #'s of the prior and new one; new one is asynch (right)?) so why wasn't it delayed; wouldn't the interrupted context still be dynamorio_sigreturn and not fcache_return; etc.

> Also, there's an NYI comment where SC_XIP is defined. Not sure what that's for.

Looks like an obsolete comment.

abhinav92003 commented:
> ```
>  Internal Error: DynamoRIO debug check failure: /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/core/unix/signal.c:6354 (app_pc)sc-SC_XIP != next_pc 
> ```
Following up on this failure in `linux.signal0111`:

It seems that the new `append_fcache_enter_prologue` code I added in this PR isn't working as intended for negative `signals_pending` vals -- should enter fcache, but isn't. Seems that the `CMP signals_pending LE 0` check is still not working.

A hang in `linux.signest` was fixed when I made `signals_pending` a `signed char` explicitly in 75b1231. But seems that that doesn't help with the comparison in assembly, which is causing the `ASSERT(sc-SC_XIP != next_pc)`  failure in `linux.signal0111` (and a few others too).

To verify this, I added a `CMP signals_pending EQ 255` check too, which fixes `linux.signal0111` (no failure in 50 runs). 

Let me figure out a good fix to push. I'll also add an `assert(signals_pending>0)` at https://github.com/DynamoRIO/dynamorio/blob/32236dc8ba480be9910e5e41cfb833949ee8cb30/core/dispatch.c#L506.

abhinav92003 commented:
> Let me figure out a good fix to push. 

I added support for `OP_ldrsb` in AArch64, and used that to load the `signals_pending` signed char in `append_fcache_enter_prologue`. That fixed it. Will commit the fix soon.

derekbruening commented:
As requested offline, I ran 3 tests 20x each.  The 2 signalnnnn ones passed all 20 times, but ptsig failed on the 10th run:

```
$ ctest --repeat-until-fail 20 -R 'signal0111|signal1111|client\.timer|ptsig'
Test project /home/derek/dr/git/build_dbg_tests
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.03 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.11 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.05 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.19 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.09 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.06 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.08 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.03 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.04 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.10 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.08 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.04 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.06 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.05 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.05 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.11 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.13 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.16 sec
    Start 35: code_api|linux.signal0111
    Test #35: code_api|linux.signal0111 ........   Passed    3.15 sec
    Start 35: code_api|linux.signal0111
1/3 Test #35: code_api|linux.signal0111 ........   Passed    3.17 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.21 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.34 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.27 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.21 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.29 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.30 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.27 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.21 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.35 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.21 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.25 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.28 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.27 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.23 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.24 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.35 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.19 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.19 sec
    Start 43: code_api|linux.signal1111
    Test #43: code_api|linux.signal1111 ........   Passed    3.28 sec
    Start 43: code_api|linux.signal1111
2/3 Test #43: code_api|linux.signal1111 ........   Passed    3.22 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.36 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.33 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.37 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.36 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.34 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.35 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.35 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.36 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........   Passed    1.30 sec
    Start 78: code_api|pthreads.ptsig
    Test #78: code_api|pthreads.ptsig ..........***Failed  Required regular expression not found.Regex=[^Estimation of pi is 3\.142425985001098
$
]  0.96 sec
```

Let me run HEAD and see whether ptsig is flaky there too.

derekbruening commented:
Here is the log for that failure:
```
78/182 Test: code_api|pthreads.ptsig
Command: "/home/derek/dr/git/build_dbg_tests/bin32/drrun" "-s" "90" "-quiet" "-debug" "-killpg" "-stderr_mask" "0xC" "-dumpcore_mask" "0" "-code_api" "--" "/home/derek/dr/git/build_dbg_tests/suite/tests/bin/pthreads.ptsig"
Directory: /home/derek/dr/git/build_dbg_tests/suite/tests
"code_api|pthreads.ptsig" start time: Oct 01 23:35 EDT
Output:
----------------------------------------------------------
<ERROR: master_signal_handler with no siginfo (i#26?): tid=18641, sig=11>
<Application /home/derek/dr/git/build_dbg_tests/suite/tests/bin/pthreads.ptsig (18640). Cannot correctly handle received signal 11 in thread 18641.>
<end of output>
Test time =   0.96 sec
----------------------------------------------------------
Test Fail Reason:
Required regular expression not found.Regex=[^Estimation of pi is 3\.142425985001098
$
]
"code_api|pthreads.ptsig" end time: Oct 01 23:35 EDT
"code_api|pthreads.ptsig" time elapsed: 00:00:00
----------------------------------------------------------
```

derekbruening commented:
Going back to 6c638197ce9e651812f33b4f31f7710d94f0cc6d in this branch, 20x of ptsig all passed.  But running 50x did hit the same #26 failure as above.  So probably that is unrelated to this PR.

derekbruening commented:
OK so I ran ptsig 100x and had 2/100 fail without this PR and 1/100 with so I think we're good.

abhinav92003 commented:
> As requested offline, I ran 3 tests 20x each. The 2 signalnnnn ones passed all 20 times, but ptsig failed on the 10th run:

Thanks a lot for checking Derek!

I just recalled there were two others that were important for this PR. 
`linux.signal_racesys`
`linux.signest`

These were part of the earlier failures I faced in this PR so it skipped my mind before :(. Please run whenever you get the chance. Thanks again.

derekbruening commented:
> I just recalled there were two others that were important for this PR.
> `linux.signal_racesys`
> `linux.signest`

For `i4271-aarch64-linux-signal-race` built DEBUG=ON:
```
$ ctest --repeat-until-fail 20 -R 'signal_racesys|signest'
...
1/2 Test #61: code_api|linux.signal_racesys ....   Passed    1.45 sec
...
2/2 Test #66: code_api|linux.signest ...........   Passed    2.66 sec

100% tests passed, 0 tests failed out of 2
```

abhinav92003 commented:
> For `i4271-aarch64-linux-signal-race` built DEBUG=ON:

Great!
